### PR TITLE
Revamp importing to enable bumping Undici, and fixing instanceof tests within the edge runtime

### PR DESCRIPTION
**Pull request by javivelasco**
This is a big iteration on how the `@edge-runtime/*` suite of packages work, done specifically to bump `undici` as a dependency.
Some context: we wanted to add WebSocket support for the edge-light runtime, and that requires bumping Undici. Unfortunately, the mechanism we used to import it previously didn't work anymore, as it uses `worker_threads`.
Thankfully, we also had a great idea on how to fix `instanceof` tests so we can avoid using unnecessary polyfills and share Uint8Arrays between the main Node.js VM and the VM we create for the edge-light runtime.
Previously, the following code didn't work:
```ts
const vm = new EdgeVM();
vm.context.MY_BINARY_ARRAY = new Uint8Array([1,2,3]);
assert(vm.evaluate(`MY_BINARY_ARRAY instanceof Uint8Array`));
assert(vm.evaluate(`MY_BINARY_ARRAY instanceof Object`));

// If the previous statements were correct, this failed:
assert(vm.evaluate(`{} instanceof Object`));
vm.context.MY_OBJ = {};
assert(vm.evaluate(`MY_OBJ instanceof Object`))
```
This is due to the fact that every V8 VM has its own `Object`, `RegExp`, `Uint8Array` (etc) classes. We fixed it now, and objects created outside of the VM will now be `instanceof` the classes from inside the VM. <picture data-single-emoji=":yesss:" title=":yesss:">:yesss:</picture> 
This little change means that instead of loading modules within the VM, we can load them in the main Node.js program, and everything will work as intended: so we can upgrade Undici!
Since we still support older Node.js versions, we need to polyfill some stuff, which requires special handling for importing things. Eventually I believe we will be able to drop them.
I also didn't do any conditional logic of imports _yet_. I believe we _can_ do that and offload to the native implementations when possible (like TextEncoder and TextDecoder, or even use the native undici implementation that exist for newer versions)
More noteworthy changes:
1. I extracted the tests from `@edge-runtime/primitives` into a new module called `@edge-runtime/integration-tests`, which, surprisingly, only runs integration tests. This fixes the circular dependency of `primitives -> jest-environment -> primitives`, which was annoying. Primitives also requires to build before running the tests so it's not like we're making a worse iteration speed. These tests are using `@edge-runtime/ponyfill` so they actually being tested both using the `@edge-runtime/jest-environment` and the standard Node.js environment using the same code!
2. Importing `@edge-runtime/primitives/*` is _discouraged_. Since there is a mandatory order of imports, and we don't use the standard `require` to import the dependencies, for good and bad.
3. Bumping `undici` means that we now have the `getSetCookies` function. We still support the `getAll` to avoid breaking changse, but internally it uses `getSetCookies`.
4. The `structuredClone` polyfill now supports `ReadableStream` (as mentioned in MDN), because Undici uses it.
5. We expose `WebSocket` constructor now! (resolves EC-617)

### CODE REVIEW

File: packages/primitives/src/primitives/fetch.js [hunk 1]
```diff
@@ -17,79 +17,33 @@ global.AbortSignal = AbortSignal
...
+const Request = new Proxy(BaseRequest, {
+  construct(target, args) {
+    const [input, init] = args
+    return new target(input, addDuplexToInit(init))
+  },
+})
+
+const __entries = HeadersModule.Headers.prototype.entries
+HeadersModule.Headers.prototype.entries = function* () {
+  for (const [key, value] of __entries.call(this)) {
```
Comment (nuta) [inline packages/primitives/src/primitives/fetch.js:31]: question: Why do we need to wrap the original `entries`? Looks like this is for `set-cookie` but would like to know the background why undici's implementation is not sufficient for us.

File: packages/primitives/package.json [hunk 2]
```diff
@@ -66,7 +65,7 @@
     "clean:build": "rm -rf dist abort-controller blob console crypto encoding events fetch streams structured-clone url",
     "clean:node": "rm -rf node_modules",
     "prebuild": "pnpm run clean:build",
-    "test": "jest"
+    "test": "#"
```
Comment (nuta) [inline packages/primitives/package.json:68]: nit: Perhaps you want to revert?

File: packages/node-utils/test/node-to-edge/request.test.ts [hunk 3]
```diff
@@ -130,7 +130,7 @@ it('does not allow to read the body twice', async () => {
 
   expect(request.method).toEqual('POST')
   expect(await request.text()).toEqual('Hello World')
-  await expect(request.text()).rejects.toThrowError('The body has already been consumed.')
+  await expect(request.text()).rejects.toThrowError('Body is unusable')
```
Comment (nuta) [inline packages/node-utils/test/node-to-edge/request.test.ts:133]: Does this new error message come from undici? The previous one looks more user-friendly to me.

File: docs/pages/packages/format.mdx [hunk 4]
```diff
@@ -8,22 +8,10 @@ It follows the [formatter spec](https://console.spec.whatwg.org/#formatter) and
...
-    yarn add @edge-runtime/format
-    ```
-  </Tab>
-  <Tab>
-    ```sh
-    pnpm install @edge-runtime/format --save
-    ```
-  </Tab>
+<Tabs items={['npm', 'yarn', 'pnpm']} storageKey='selected-pkg-manager'>
+  <Tab>```sh npm install @edge-runtime/format --save ```</Tab>
```
Comment (nuta) [inline docs/pages/packages/format.mdx:12]: This code block will look like following (`sh` is not recognized as language name!):

```sh npm install @edge-runtime/format --save ```

IIRC formatter does this broken formatting.

File: docs/pages/packages/node-utils.mdx [hunk 5]
```diff
@@ -11,22 +11,10 @@ The **@edge-runtime/node-utils** package contains utilities to run web compliant
 
 ## Installation
 
-<Tabs items={['npm', 'yarn', 'pnpm']} storageKey="selected-pkg-manager">
```
Comment (nuta) [inline docs/pages/packages/node-utils.mdx:14]: ditto

File: docs/pages/packages/ponyfill.mdx [hunk 6]
```diff
@@ -11,22 +11,10 @@ Note this is just necessary if you want to run the same code across different en
 
 ## Installation
 
-<Tabs items={['npm', 'yarn', 'pnpm']} storageKey="selected-pkg-manager">
```
Comment (nuta) [inline docs/pages/packages/ponyfill.mdx:14]: ditto

File: docs/pages/packages/primitives.mdx [hunk 7]
```diff
@@ -10,22 +10,10 @@ See full list is available at [Edge Runtime APIs](/features/available-apis).
 
 ## Installation
 
-<Tabs items={['npm', 'yarn', 'pnpm']} storageKey="selected-pkg-manager">
```
Comment (nuta) [inline docs/pages/packages/primitives.mdx:13]: ditto

File: packages/vm/src/edge-vm.ts [hunk 8]
```diff
@@ -72,12 +76,52 @@ export class EdgeVM<T extends EdgeContext = EdgeContext> extends VM<T> {
 
     this.evaluate<void>(getDefineEventListenersCode())
     this.dispatchFetch = this.evaluate<DispatchFetch>(getDispatchFetchCode())
+    for (const item of [
```
Comment (nuta) [inline packages/vm/src/edge-vm.ts:79]: Do you mind adding a brief comment why we need to patch these specific ones? 🙏

**Review by nuta — COMMENTED** [review]
Nice! 💯 Left some nitpicks for now. Will look at the details later.

File: packages/primitives/package.json (hunk 2 above)
Comment (Schniz) [reply packages/primitives/package.json:68]: I need to remove the `test` script altogether
it is now tested in `packages/integration-tests`, as the previous tests were depending on `jest-environment` and created a circular dependency loop, so `pnpm run test` failed without building in advance <picture data-single-emoji=":wow:" title=":wow:">:wow:</picture>

File: packages/node-utils/test/node-to-edge/request.test.ts (hunk 3 above)
Comment (Schniz) [reply packages/node-utils/test/node-to-edge/request.test.ts:133]: Yeah, this is from undici afaik. I can maybe map it but I'm not sure I want to <picture data-single-emoji=":sherthonk_holmes:" title=":sherthonk_holmes:">:sherthonk_holmes:</picture> wdyt?

File: docs/pages/packages/format.mdx (hunk 4 above)
Comment (Schniz) [reply docs/pages/packages/format.mdx:12]: ah right will revert all changes to `docs/pages/**/*.mdx` <picture data-single-emoji=":relieved-telegram:" title=":relieved-telegram:">:relieved-telegram:</picture>

File: packages/vm/src/edge-vm.ts (hunk 8 above)
Comment (Schniz) [reply packages/vm/src/edge-vm.ts]: <picture data-single-emoji=":done:" title=":done:">:done:</picture>

File: docs/pages/packages/primitives.mdx (hunk 7 above)
Comment (Schniz) [reply docs/pages/packages/primitives.mdx]: <picture data-single-emoji=":done:" title=":done:">:done:</picture>

File: docs/pages/packages/ponyfill.mdx (hunk 6 above)
Comment (Schniz) [reply docs/pages/packages/ponyfill.mdx]: <picture data-single-emoji=":ditto:" title=":ditto:">:ditto:</picture>

File: docs/pages/packages/node-utils.mdx (hunk 5 above)
Comment (Schniz) [reply docs/pages/packages/node-utils.mdx]: <picture data-single-emoji=":ditto-dance:" title=":ditto-dance:">:ditto-dance:</picture>

File: docs/pages/packages/format.mdx (hunk 4 above)
Comment (Schniz) [reply docs/pages/packages/format.mdx]: <picture data-single-emoji=":ditto:" title=":ditto:">:ditto:</picture>

File: packages/primitives/src/primitives/fetch.js [hunk 9]
```diff
@@ -17,79 +17,33 @@ global.AbortSignal = AbortSignal
...
-HeadersModule.HeadersList.prototype.append = function (name, value) {
-  const result = __append.call(this, name, value)
-  if (!this[SCookies]) {
-    Object.defineProperty(this, SCookies, {
-      configurable: false,
-      enumerable: false,
-      writable: true,
-      value: [],
-    })
+process.emitWarning = () => {}
```
Comment (Kikobeats) [inline packages/primitives/src/primitives/fetch.js:20]:
```suggestion
process.emitWarning = void
```

**Review by Kikobeats — CHANGES_REQUESTED** [review]

**Review by Kikobeats — APPROVED** [review]

**Pull request merged**